File: android/app/src/main/java/com/magicforms/mobile/data/api/SearchApi.kt
package com.magicforms.mobile.data.api

import com.magicforms.mobile.data.models.SubmissionSearchResponse
import kotlinx.serialization.json.Json

class SearchApi {
    private val json = Json { ignoreUnknownKeys = true; isLenient = true }
    private val authApi = AuthApi()

    fun fetchSubmissions(
        token: String,
        query: String = "",
        relation: String = "any",
        workflowState: String? = null,
        formId: Int? = null,
        limit: Int = 50,
        offset: Int = 0,
    ): SubmissionSearchResponse {
        val params = buildList {
            add("limit=$limit")
            add("offset=$offset")
            add("relation=$relation")
            if (query.isNotBlank()) add("q=${java.net.URLEncoder.encode(query.trim(), Charsets.UTF_8.name())}")
            if (!workflowState.isNullOrBlank()) add("workflow_state=$workflowState")
            if (formId != null) add("form=$formId")
        }
        val url = ApiConfig.url("/api/v1/submissions/search/?${params.joinToString("&")}")
        val data = authApi.get(url, token)
        return json.decodeFromString(data)
    }
}

File: ios/MagicFormsMobile/Services/SearchAPI.swift
import Foundation

enum SearchAPI {
    private static let decoder = JSONDecoder()

    static func fetchSubmissions(
        token: String,
        query: String = "",
        relation: String = "any",
        workflowState: String? = nil,
        formId: Int? = nil,
        limit: Int = 50,
        offset: Int = 0
    ) async throws -> SubmissionSearchResponse {
        var components = URLComponents(url: APIConfig.url(path: "/api/v1/submissions/search/"), resolvingAgainstBaseURL: false)!
        var queryItems = [
            URLQueryItem(name: "limit", value: String(limit)),
            URLQueryItem(name: "offset", value: String(offset)),
            URLQueryItem(name: "relation", value: relation),
        ]
        let trimmed = query.trimmingCharacters(in: .whitespacesAndNewlines)
        if !trimmed.isEmpty {
            queryItems.append(URLQueryItem(name: "q", value: trimmed))
        }
        if let workflowState, !workflowState.isEmpty {
            queryItems.append(URLQueryItem(name: "workflow_state", value: workflowState))
        }
        if let formId {
            queryItems.append(URLQueryItem(name: "form", value: String(formId)))
        }
        components.queryItems = queryItems

        let data = try await AuthAPI.getData(url: components.url!, token: token)
        let response = try decoder.decode(SubmissionSearchResponse.self, from: data)
        if response.ok { return response }
        throw try AuthAPI.decodeError(from: data)
    }
}
